zerver/actions/message_send.py

```diff
@@ -313,11 +339,11 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
         #         to-do.
         rows = []
 
-    def get_ids_for(f: Callable[[Dict[str, Any]], bool]) -> Set[int]:
+    def get_ids_for(f: Callable[[ActiveUserDict], bool]) -> Set[int]:
         """Only includes users on the explicit message to line"""
         return {row["id"] for row in rows if f(row)} & message_to_user_id_set
 
-    def is_service_bot(row: Dict[str, Any]) -> bool:
+    def is_service_bot(row: ActiveUserDict) -> TypeGuard[ActiveBotUserDict]:
         return row["is_bot"] and (row["bot_type"] in UserProfile.SERVICE_BOT_TYPES)
```

Comment:
Since the type safety of a `TypeGuard` is unchecked by mypy, I’d somewhat prefer to avoid them where there are reasonable alternatives: #22498.

Reply:
Yes. I imagine that a more appropriate use case will be where we have yet unavoidable `cast`. It wouldn't make a huge difference in terms of type safety, but can be more readable depending on the context.